### PR TITLE
Enable resoure configuration via Kruize CR

### PR DESCRIPTION
**Pull request by shreyabiradar07**
This PR makes CPU, memory, and storage resources configurable via the Kruize CR instead of being hardcoded in the operator.

## Summary by Sourcery

Make Kruize CPU, memory, and storage resources configurable via the Kruize custom resource instead of being hardcoded, while preserving sensible defaults.

New Features:
- Introduce a ResourceConfig section on the Kruize CRD to configure CPU, memory, and persistent volume settings for the Kruize and database components.
- Allow the resource generator to consume per-component resource configuration from the Kruize CR and apply it to Deployments and persistent storage objects across OpenShift and Kubernetes clusters.

Enhancements:
- Centralize default CPU, memory, and PV/PVC settings into constants and use helper methods in the resource generator to compute effective resources with fallback behavior.
- Extend the resource generator to accept a context and log when invalid resource quantities are provided, falling back to defaults.
- Refactor tests to use shared helpers and cover default, custom, and partial resource configuration for Deployments, PVs, PVCs, Pods, and Services.

Documentation:
- Update the sample Kruize custom resource to document optional resource and persistent volume configuration fields.

Tests:
- Add comprehensive unit tests validating custom and default resource behavior for Kruize and database Deployments, including partial overrides and error handling for invalid quantities.
- Add tests verifying PV and PVC sizing, storage class, host path, and access modes when configured via the CRD and when using defaults on both OpenShift and Kubernetes.

### CODE REVIEW

File: internal/controller/kruize_controller_test.go [hunk 1]
```diff
@@ -298,39 +298,39 @@ var _ = Describe("Kruize Controller", func() {
 
 	Context("Resource generation", func() {
 		It("should generate cluster-scoped resources for OpenShift", func() {
```
[the remark this answers is not in the document]
Comment (shreyabiradar07) [reply internal/controller/kruize_controller_test.go:461]: Added testcase for custom ResourceConfig

File: internal/controller/kruize_controller_test.go [hunk 2]
```diff
@@ -298,39 +298,39 @@ var _ = Describe("Kruize Controller", func() {
...
 		It("should generate namespaced resources for OpenShift", func() {
-			generator := utils.NewKruizeResourceGenerator("test-namespace", "", "", constants.ClusterTypeOpenShift)
+			generator := utils.NewKruizeResourceGenerator("test-namespace", "", "", constants.ClusterTypeOpenShift, nil)
 
 			namespacedResources := generator.NamespacedResources()
 			Expect(namespacedResources).NotTo(BeEmpty())
 			Expect(len(namespacedResources)).To(BeNumerically(">", 0))
 		})
 
 		It("should generate Kubernetes cluster-scoped resources", func() {
```
[the remark this answers is not in the document]
Comment (shreyabiradar07) [reply internal/controller/kruize_controller_test.go:477]: Added testcase for custom PV/PVC configuration for Openshift and  Kubernetes

File: internal/controller/kruize_controller_test.go [hunk 3]
```diff
@@ -503,16 +503,16 @@ var _ = Describe("Kruize Controller", func() {
 					break
 				}
 			}
-			
+
 			Expect(kruizeUIPod).NotTo(BeNil(), "Kruize UI pod should exist")
 			Expect(kruizeUIPod.Spec.Containers).NotTo(BeEmpty())
 		})
 
 		It("should generate Kruize-db pod specification", func() {
```
[the remark this answers is not in the document]
Comment (shreyabiradar07) [reply internal/controller/kruize_controller_test.go:789]: Added edge-case test for partial ResourceConfig

**Comment by shreyabiradar07** [comment]
@sourcery-ai review

**Comment by shreyabiradar07** [comment]
@sourcery-ai review

**Comment by shreyabiradar07** [comment]
@sourcery-ai review

**Comment by shreyabiradar07** [comment]
@bharathappali Updated the Kruize CR resource config hierarchy as discussed in the design call, please review 

Current values are similar as [kruize-crc-openshift.yaml](https://github.com/kruize/autotune/blob/master/manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml) as openshift is the default.

**Note**: PV and PVC are independent k8s resources, hence did not nest it under kruize-db, added persistentVolumeClaim under volumes for kruize-db

`v1alpha1_kruize.yaml:`
```
apiVersion: kruize.io/v1alpha1
kind: Kruize
metadata:
  labels:
    app.kubernetes.io/name: kruize-operator
    app.kubernetes.io/managed-by: kustomize
  name: kruize-sample
spec:
  # Supported values: "openshift", "minikube", "kind"
  cluster_type: "openshift"
  autotune_image: "quay.io/kruize/autotune_operator:0.9"
  autotune_ui_image: "quay.io/kruize/kruize-ui:0.1.0"
  # For OpenShift, use: "openshift-tuning", Minikube/KIND, use: "monitoring"
  namespace: "openshift-tuning"
  
  # Optional: Resource configuration for Kruize components
  # If not specified, default values will be used
  # Persistent Volume configuration
  persistentVolume:
    name: "kruize-db-pv-volume"
    storageClassName: "manual"
    capacity:
      storage: "500Mi"
    accessModes:
      - "ReadWriteMany"
    hostPath:
      path: "/mnt/data"
    labels:
      type: "local"
      app: "kruize-db"
  
  # Persistent Volume Claim configuration
  persistentVolumeClaim:
    name: "kruize-db-pv-claim"
    storageClassName: "manual"
    accessModes:
      - "ReadWriteMany"
    resources:
      requests:
        storage: "500Mi"
    labels:
      app: "kruize-db"
  
  # Database resource configuration
  kruize-db:
    resources:
      requests:
        memory: "100Mi"
        cpu: "0.5"
      limits:
        memory: "100Mi"
        cpu: "0.5"
    
    # Volume mount configuration (container-level)
    volumeMounts:
      - name: "kruize-db-storage"
        mountPath: "/var/lib/pgsql/data"
    
    # Volume configuration (pod-level)
    volumes:
      - name: "kruize-db-storage"
        persistentVolumeClaim:
          claimName: "kruize-db-pv-claim"
  
  # Kruize application resource configuration
  kruize:
    resources:
      requests:
        memory: "768Mi"
        cpu: "0.7"
      limits:
        memory: "768Mi"
        cpu: "0.7"
```

**Review by bharathappali — APPROVED** [review]
LGTM

File: internal/constants/resource_defaults.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// Kruize application resource defaults
+const (
+	// DefaultKruizeCPURequest is the default CPU request for Kruize application
+	DefaultKruizeCPURequest = "0.7"
+	// DefaultKruizeCPULimit is the default CPU limit for Kruize application
+	DefaultKruizeCPULimit = "0.7"
+	// DefaultKruizeMemoryRequest is the default memory request for Kruize application
+	DefaultKruizeMemoryRequest = "768Mi"
+	// DefaultKruizeMemoryLimit is the default memory limit for Kruize application
+	DefaultKruizeMemoryLimit = "768Mi"
```
Comment (dinogun) [inline internal/constants/resource_defaults.go:32]: Is this applicable for all K8s targets (OCP, Kind, minikube)? Currently we dont set req / lim for minikube / kind as that can prevent folks with small config laptops to use Kruize

Comment (shreyabiradar07) [reply internal/constants/resource_defaults.go:32]: Currently, these defaults serve as a fallback for all K8s targets only when the custom resource(CR) is missing resource specs. 
While they are fully configurable via the CR, I understand the concern for users on smaller laptops and will disable the defaults for Kind and Minikube

Comment (shreyabiradar07) [reply internal/constants/resource_defaults.go:32]: Updated to apply default resources only for OpenShift

**Review by dinogun — APPROVED** [review]
LGTM
